drivers/mil_passive_sonar/passive_sonar_driver/passive_sonar_driver.py

```diff
@@ -0,0 +1,590 @@
+                    .format(type(custom_input_source), 'ReceiverArrayInterface'))
+            self.input_source = custom_input_source
+
+        else:
+            raise RuntimeError(self.input_mode + 'is not a supported input mode')
+
+        self.reset_position_estimate(None)
+
+        self.multilaterator = Multilaterator(self.receiver_locations, self.c, self.method)
+
+        self.plot_pub = rospy.Publisher('/passive_sonar/plot', Image, queue_size=1)
+        self.rviz_pub = rospy.Publisher("/passive_sonar/rviz", Marker, queue_size=10)
+        self.declare_services()
+        rospy.loginfo('Passive sonar driver initialized')
+
+    # Passive Sonar Helpers
+
+    def load_params(self):
+        '''
+        Loads all the parameters needed for receiving and processing signals from the passive
+        sonar board and calculating headings towards an active pinger.
+
+        These parameters are descrived in detail in the Passive Sonar page of the mil_common wiki.
+        TODO: copy url here
+        '''
+        # ROS params expected to be loaded under namespace passive_sonar
+        self.required_params = ['receiver_locations', 'method', 'c', 'target_frequency',
+                                'sampling_frequency', 'upsampling_factor', 'locating_frame',
+                                'receiver_array_frame', 'min_variance', 'observation_buffer_size',
+                                'input_timeout']
+        self.input_src_params = \
+        {
+         'serial' : ['port', 'baud', 'tx_request_code', 'tx_start_code', 'read_timeout',
+                     'scalar_size', 'signal_size', 'signal_bias', 'locating_frame',
+                     'receiver_array_frame'],
+         'bag'    : ['bag_filename' ,'locating_frame', 'receiver_array_frame']
+        }
+
+        load = lambda prop: setattr(self, prop, rospy.get_param('passive_sonar/' + prop))
+        try:
+            [load(x) for x in self.required_params]
+        except KeyError as e:
+            raise IOError('A required rosparam was not declared: ' + str(e))
+
+        self.receiver_count = len(self.receiver_locations) + 1
+        self.receiver_locations = np.array(  # dictionary to numpy array
+            [np.array([x['x'], x['y'], x['z']]) for x in self.receiver_locations])
+
+    def declare_services(self):
+       '''
+       Conveniently declares all the services offered by the driver
+       '''
+       services = {
+                      'get_pulse_heading'        : GetPulseHeading,
+                      'estimate_pinger_position' : EstimatePingerPosition,
+                      'reset_position_estimate'  : ResetPositionEstimate,
+                      'start_bagging'            : StartBagging,
+                      'save_bag'                 : SaveBag,
+                      'set_frequency'            : SetFrequency
+                  }
+       [rospy.Service('passive_sonar/' + s[0], s[1], getattr(self, s[0])) for s in services.items()]
```

Comment:
Do the services stay alive when you don't assign them to a variable? I feel like I've had issues where the service wouldn't be available in the past when I don't set the instantiation to a variable - thought I could have been a fool in those days.